qiskit/providers/ibmq/job/ibmqjob.py

```diff
@@ -451,3 +481,23 @@ def _wait_for_completion(
             self.refresh()
 
         return self._status in required_status
+
+    def _retrieve_result(self) -> Optional[Result]:
+        """Retrieve the job result response.
+
+        Returns:
+            The job result, or ``None`` if schema validation fails.
+
+        Raises:
+            IBMQJobApiError: if there was some unexpected failure in the server.
+        """
+        if not self._result_response:
+            with api_to_job_error():
+                self._result_response = self._api.job_result(self.job_id(),
+                                                             self._use_object_storage)
+            try:
+                return Result.from_dict(self._result_response)
```

Comment:
If you set `self._result` here instead of returning it, then you can consolidate lines 249-256 to just `return self._retrieve_result()`. It would also keep the api call to retrieve result to a single method, making the code cleaner.